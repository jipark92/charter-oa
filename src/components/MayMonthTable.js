import React from 'react'

export default function MayMonthTable({ sortedMayTable, calculateMayTotalRewards }) {
    return (
        <div>
            <h1 data-testid='header-may'>MAY</h1>

            {/* customer 123 */}
            <table border="1"  >
                <thead >
                    <tr>
                        <th data-testid='th-customerid'>Customer ID</th>
                        <th data-testid='th-transactionid'>Transaction ID</th>
                        <th data-testid='th-date'>Date</th>
                        <th data-testid='th-price'>Price</th>
                        <th data-testid='th-rewards'>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedMayTable().map((data, i) => {
                        if (data.customerID === 123) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td data-testid='customer123Total'><b><i>{calculateMayTotalRewards(123)}</i></b></td>
                    </tr>
                </tbody>
            </table>

            {/* customer 456 */}
            <table border="1"  >
                <thead>
                    <tr>
                        <th data-testid='th-customerid'>Customer ID</th>
                        <th data-testid='th-transactionid'>Transaction ID</th>
                        <th data-testid='th-date'>Date</th>
                        <th data-testid='th-price'>Price</th>
                        <th data-testid='th-rewards'>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedMayTable().map((data, i) => {
                        if (data.customerID === 456) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td data-testid='customer456Total'><b><i>{calculateMayTotalRewards(456)}</i></b></td>
                    </tr>
                </tbody>
            </table>

            {/* customer 789 */}
            <table border="1"  >
                <thead>
                    <tr>
                        <th data-testid='th-customerid'>Customer ID</th>
                        <th data-testid='th-transactionid'>Transaction ID</th>
                        <th data-testid='th-date'>Date</th>
                        <th data-testid='th-price'>Price</th>
                        <th data-testid='th-rewards'>Rewards</th>
                    </tr>
                </thead>
                <tbody>
                    {sortedMayTable().map((data, i) => {
                        if (data.customerID === 789) {
                            return (
                                <tr key={i}>
                                    <td>{data.customerID}</td>
                                    <td>{data.transactionID}</td>
                                    <td>{data.purchasedDate}</td>
                                    <td>${data.price}</td>
                                    <td>{data.rewardPoints}</td>
                                </tr>
                            )
                        }
                    })}
                    <tr>
                        <td colSpan="4"><b>TOTAL REWARDS</b></td>
                        <td data-testid='customer789Total'><b><i>{calculateMayTotalRewards(789)}</i></b></td>
                    </tr>
                </tbody>
            </table>
        </div >
    )
}