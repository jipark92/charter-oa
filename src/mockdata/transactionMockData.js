export const calculateRewardPoints = (price) => {
    const moreThan50 = 1
    const moreThan100 = 2
    if (price > 100) {
        return (price - 100) * moreThan100 + 50 * moreThan50
    } else if (price > 50) {
        return (price - 50) * moreThan50
    } else {
        return 0
    }
}

const transactionMockData = [
    {
        transactionID: 1,
        customerID: 123,
        price: 100,
        rewardPoints: calculateRewardPoints(100),
        purchasedDate: new Date("July 01, 22 ").toDateString(),
    },
    {
        transactionID: 2,
        customerID: 456,
        price: 60,
        rewardPoints: calculateRewardPoints(60),
        purchasedDate: new Date("June 19, 22 ").toDateString(),
    },
    {
        transactionID: 3,
        customerID: 789,
        price: 175,
        rewardPoints: calculateRewardPoints(175),
        purchasedDate: new Date("May 23, 22 ").toDateString(),
    },
    {
        transactionID: 4,
        customerID: 123,
        price: 55,
        rewardPoints: calculateRewardPoints(55),
        purchasedDate: new Date("May 21, 22 ").toDateString(),
    },
    {
        transactionID: 5,
        customerID: 456,
        price: 125,
        rewardPoints: calculateRewardPoints(125),
        purchasedDate: new Date("May 11, 22 ").toDateString(),
    },
    //
    {
        transactionID: 6,
        customerID: 789,
        price: 100,
        rewardPoints: calculateRewardPoints(100),
        purchasedDate: new Date("June 25, 22 ").toDateString(),
    },
    {
        transactionID: 7,
        customerID: 123,
        price: 45,
        rewardPoints: calculateRewardPoints(45),
        purchasedDate: new Date("June 14, 22 ").toDateString(),
    },
    {
        transactionID: 8,
        customerID: 456,
        price: 203,
        rewardPoints: calculateRewardPoints(203),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 9,
        customerID: 789,
        price: 126,
        rewardPoints: calculateRewardPoints(126),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 10,
        customerID: 123,
        price: 500,
        rewardPoints: calculateRewardPoints(500),
        purchasedDate: new Date("May 14, 22 ").toDateString(),
    },
    {
        transactionID: 11,
        customerID: 456,
        price: 75,
        rewardPoints: calculateRewardPoints(75),
        purchasedDate: new Date("June 15, 22 ").toDateString(),
    },
    {
        transactionID: 12,
        customerID: 789,
        price: 89,
        rewardPoints: calculateRewardPoints(89),
        purchasedDate: new Date("July 15, 22 ").toDateString(),
    },
    {
        transactionID: 13,
        customerID: 123,
        price: 152,
        rewardPoints: calculateRewardPoints(152),
        purchasedDate: new Date("May 15, 22 ").toDateString(),
    },
];

export const fetchTransactionMockData = () => {
    return new Promise((res, rej) => {
        setTimeout(() => {
            res(transactionMockData)
        }, 1000);
    })
}