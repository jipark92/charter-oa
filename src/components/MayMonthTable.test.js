import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MayMonthTable from "./MayMonthTable";
import { transactionMockData } from '../mockdata/transactionMockData'


//categorized table array
const mayTable = []

//find may data
const findMayDateAndCustomerID = (() => {
    transactionMockData?.map((data) => {
        if (data.customerID && new Date(data.purchasedDate).getMonth() + 1 === 5) {
            mayTable.push(data)
        }
    })
})()
const sortedMayTable = () => mayTable.sort((a, b) => new Date(a.purchasedDate) - new Date(b.purchasedDate))

//calcualte may total reward points
const calculateMayTotalRewards = (customer) => {
    const mayRewardPointsArr = []
    mayTable.map((data) => {
        if (data.customerID === customer) {
            mayRewardPointsArr.push(data.rewardPoints)
        }
    })
    return mayRewardPointsArr.reduce((total, pointsArr) => {
        return total + pointsArr
    }, 0)
}

//test
describe(`MayMonthTable should render customer's May information in tables`, () => {

    it('should match snapshot', () => {
        const { asFragment } = render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('header should be MAY', () => {
        render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)
        const headerMay = screen.getByTestId('header-may').textContent
        expect(headerMay).toBe('MAY')
    })

    it('there should be 3 TableHeaders in May because there are 3 customers', () => {
        render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)

        expect(screen.getAllByTestId('th-customerid').length).toEqual(3)
        expect(screen.getAllByTestId('th-transactionid').length).toEqual(3)
        expect(screen.getAllByTestId('th-date').length).toEqual(3)
        expect(screen.getAllByTestId('th-price').length).toEqual(3)
        expect(screen.getAllByTestId('th-rewards').length).toEqual(3)
    })

    it('total for customer 123 should be 1009', () => {
        render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)
        const customer123Total = screen.getByTestId('customer123Total').textContent
        expect(customer123Total).toBe('1009')
    })

    it('total for customer 456 should be 100', () => {
        render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)
        const customer456Total = screen.getByTestId('customer456Total').textContent
        expect(customer456Total).toBe('100')
    })

    it('total for customer 789 should be 2512', () => {
        render(<MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
        />)
        const customer789Total = screen.getByTestId('customer789Total').textContent
        expect(customer789Total).toBe('2512')
    })

})