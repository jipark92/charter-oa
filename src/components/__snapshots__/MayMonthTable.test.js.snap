// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MayMonthTable should render customer's May information in tables should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      data-testid="header-may"
    >
      MAY
    </h1>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            123
          </td>
          <td>
            13
          </td>
          <td>
            Fri May 13 2022
          </td>
          <td>
            $152
          </td>
          <td>
            154
          </td>
        </tr>
        <tr>
          <td>
            123
          </td>
          <td>
            4
          </td>
          <td>
            Sat May 21 2022
          </td>
          <td>
            $55
          </td>
          <td>
            5
          </td>
        </tr>
        <tr>
          <td>
            123
          </td>
          <td>
            10
          </td>
          <td>
            Tue May 24 2022
          </td>
          <td>
            $500
          </td>
          <td>
            850
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer123Total"
          >
            <b>
              <i>
                1009
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            456
          </td>
          <td>
            5
          </td>
          <td>
            Wed May 11 2022
          </td>
          <td>
            $125
          </td>
          <td>
            100
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer456Total"
          >
            <b>
              <i>
                100
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
    <table
      border="1"
    >
      <thead>
        <tr>
          <th
            data-testid="th-customerid"
          >
            Customer ID
          </th>
          <th
            data-testid="th-transactionid"
          >
            Transaction ID
          </th>
          <th
            data-testid="th-date"
          >
            Date
          </th>
          <th
            data-testid="th-price"
          >
            Price
          </th>
          <th
            data-testid="th-rewards"
          >
            Rewards
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            789
          </td>
          <td>
            14
          </td>
          <td>
            Fri May 13 2022
          </td>
          <td>
            $1231
          </td>
          <td>
            2312
          </td>
        </tr>
        <tr>
          <td>
            789
          </td>
          <td>
            3
          </td>
          <td>
            Mon May 23 2022
          </td>
          <td>
            $175
          </td>
          <td>
            200
          </td>
        </tr>
        <tr>
          <td
            colspan="4"
          >
            <b>
              TOTAL REWARDS
            </b>
          </td>
          <td
            data-testid="customer789Total"
          >
            <b>
              <i>
                2512
              </i>
            </b>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
